Sources/SwiftFrameCore/Workers/ConfigProcessor.swift

```diff
@@ -0,0 +1,114 @@
+import AppKit
+import Foundation
+
+public class ConfigProcessor {
+
+    // MARK: - Properties
+
+    private var data: ConfigData
+    var verbose: Bool
+
+    // MARK: - Init
+
+    public init(filePath: String, verbose: Bool) throws {
+        let configURL = URL(fileURLWithPath: filePath)
+        let bytes = try Data(contentsOf: configURL)
+
+        data = try JSONDecoder().decode(ConfigData.self, from: bytes)
+        self.verbose = verbose
+    }
+
+    // MARK: - Methods
+
+    public func validate() throws {
+        try process()
+        try data.validate()
+    }
+
+    private func process() throws {
+        try data.process()
+    }
+
+    public func run() throws {
+        if verbose {
+            data.printSummary(insetByTabs: 0)
+            print("Press return key to continue")
+            _ = readLine()
+        }
+
+        print("Parsed and validated config file\nRendering...\n")
+
+        var fileWriteFinishers = 0
+        let requiredFileWrites = data.deviceData.count * data.titles.count
```

Comment:
That was my initial approach, which resulted in a deadlock since the dispatch group blocks the main thread - which is being used by the string parsing

Reply:
Right, the dispatch group cannot wait on the main thread. However, I think what you could do is "block" the main thread with the run loop semaphore like you do now, then dispatch to a background queue, create the dispatch group there and then dispatch the individual tasks from there.

I think what you have now should work in most cases. However, I believe things like `fileWriteFinishers += 1` are not atomic operations (it's executed as `let tmp = fileWriteFinishers + 1; fileWriteFinishers = tmp`) which means with multiple threads there's the potential for race conditions that lead to incorrect values. Dispatch groups and semaphore are generally made exactly to avoid such manual counting and the potential problems associated with it.

Reply:
Did you try moving the dispatch group to a background thread?